File: PolicyPlus/OpenPol.vb
Public Class OpenPol
    Public SelectedUser, SelectedComputer As PolicyLoader
    Public Sub SetLastSources(ComputerType As PolicyLoaderSource, ComputerData As String, UserType As PolicyLoaderSource, UserData As String)
        ' Set up the UI to show the given configuration
        Select Case ComputerType
            Case PolicyLoaderSource.LocalGpo
                CompLocalOption.Checked = True
            Case PolicyLoaderSource.LocalRegistry
                CompRegistryOption.Checked = True
                CompRegTextbox.Text = ComputerData
            Case PolicyLoaderSource.PolFile
                CompFileOption.Checked = True
                CompPolFilenameTextbox.Text = ComputerData
            Case PolicyLoaderSource.Null
                CompNullOption.Checked = True
        End Select
        Select Case UserType
            Case PolicyLoaderSource.LocalGpo
                UserLocalOption.Checked = True
            Case PolicyLoaderSource.LocalRegistry
                UserRegistryOption.Checked = True
                UserRegTextbox.Text = UserData
            Case PolicyLoaderSource.PolFile
                UserFileOption.Checked = True
                UserPolFilenameTextbox.Text = UserData
            Case PolicyLoaderSource.SidGpo
                UserPerUserGpoOption.Checked = True
                UserGpoSidTextbox.Text = UserData
            Case PolicyLoaderSource.NtUserDat
                UserPerUserRegOption.Checked = True
                UserHivePathTextbox.Text = UserData
            Case PolicyLoaderSource.Null
                UserNullOption.Checked = True
        End Select
    End Sub
    Sub BrowseForPol(DestTextbox As TextBox)
        ' Browse for a POL file and put it in a text box
        Using sfd As New SaveFileDialog
            sfd.OverwritePrompt = False
            sfd.Filter = "Registry policy files|*.pol"
            If sfd.ShowDialog = DialogResult.OK Then DestTextbox.Text = sfd.FileName
        End Using
    End Sub
    Private Sub CompOptionsCheckedChanged(sender As Object, e As EventArgs) Handles CompLocalOption.CheckedChanged, CompRegistryOption.CheckedChanged, CompFileOption.CheckedChanged, CompNullOption.CheckedChanged
        ' When the user changes the computer selection
        Dim regMount = CompRegistryOption.Checked
        CompRegTextbox.Enabled = regMount
        Dim polActive = CompFileOption.Checked
        CompPolFilenameTextbox.Enabled = polActive
        CompFileBrowseButton.Enabled = polActive
    End Sub
    Private Sub UserOptionsCheckedChanged(sender As Object, e As EventArgs) Handles UserLocalOption.CheckedChanged, UserRegistryOption.CheckedChanged, UserFileOption.CheckedChanged, UserPerUserGpoOption.CheckedChanged, UserPerUserRegOption.CheckedChanged, UserNullOption.CheckedChanged
        ' When the user changes the user selection
        Dim regMount = UserRegistryOption.Checked
        UserRegTextbox.Enabled = regMount
        Dim file = UserFileOption.Checked
        UserPolFilenameTextbox.Enabled = file
        UserFileBrowseButton.Enabled = file
        Dim perUserGpo = UserPerUserGpoOption.Checked
        UserGpoSidTextbox.Enabled = perUserGpo
        UserBrowseGpoButton.Enabled = perUserGpo
        Dim perUserHive = UserPerUserRegOption.Checked
        UserHivePathTextbox.Enabled = perUserHive
        UserBrowseHiveButton.Enabled = perUserHive
    End Sub
    Private Sub CompFileBrowseButton_Click(sender As Object, e As EventArgs) Handles CompFileBrowseButton.Click
        BrowseForPol(CompPolFilenameTextbox)
    End Sub
    Private Sub UserFileBrowseButton_Click(sender As Object, e As EventArgs) Handles UserFileBrowseButton.Click
        BrowseForPol(UserPolFilenameTextbox)
    End Sub
    Private Sub UserBrowseRegistryButton_Click(sender As Object, e As EventArgs) Handles UserBrowseHiveButton.Click
        ' Browse for a user Registry hive
        If OpenUserRegistry.ShowDialog = DialogResult.OK Then UserHivePathTextbox.Text = OpenUserRegistry.SelectedFilePath
    End Sub
    Private Sub OkButton_Click(sender As Object, e As EventArgs) Handles OkButton.Click
        ' Create the policy loaders
        Try
            If CompLocalOption.Checked Then
                SelectedComputer = New PolicyLoader(PolicyLoaderSource.LocalGpo, "", False)
            ElseIf CompRegistryOption.Checked Then
                SelectedComputer = New PolicyLoader(PolicyLoaderSource.LocalRegistry, CompRegTextbox.Text, False)
            ElseIf CompFileOption.Checked Then
                SelectedComputer = New PolicyLoader(PolicyLoaderSource.PolFile, CompPolFilenameTextbox.Text, False)
            Else
                SelectedComputer = New PolicyLoader(PolicyLoaderSource.Null, "", False)
            End If
        Catch ex As Exception
            MsgBox("The computer policy loader could not be created. " & ex.Message, MsgBoxStyle.Exclamation)
            Exit Sub
        End Try
        Try
            If UserLocalOption.Checked Then
                SelectedUser = New PolicyLoader(PolicyLoaderSource.LocalGpo, "", True)
            ElseIf UserRegistryOption.Checked Then
                SelectedUser = New PolicyLoader(PolicyLoaderSource.LocalRegistry, UserRegTextbox.Text, True)
            ElseIf UserFileOption.Checked Then
                SelectedUser = New PolicyLoader(PolicyLoaderSource.PolFile, UserPolFilenameTextbox.Text, True)
            ElseIf UserPerUserGpoOption.Checked Then
                SelectedUser = New PolicyLoader(PolicyLoaderSource.SidGpo, UserGpoSidTextbox.Text, True)
            ElseIf UserPerUserRegOption.Checked Then
                SelectedUser = New PolicyLoader(PolicyLoaderSource.NtUserDat, UserHivePathTextbox.Text, True)
            Else
                SelectedUser = New PolicyLoader(PolicyLoaderSource.Null, "", True)
            End If
        Catch ex As Exception
            MsgBox("The user policy loader could not be created. " & ex.Message, MsgBoxStyle.Exclamation)
            Exit Sub
        End Try
        DialogResult = DialogResult.OK
    End Sub
    Private Sub UserBrowseGpoButton_Click(sender As Object, e As EventArgs) Handles UserBrowseGpoButton.Click
        ' Browse for a per-user GPO
        If OpenUserGpo.ShowDialog = DialogResult.OK Then UserGpoSidTextbox.Text = OpenUserGpo.SelectedSid
    End Sub
    Private Sub OpenPol_KeyDown(sender As Object, e As KeyEventArgs) Handles Me.KeyDown
        If e.KeyCode = Keys.Escape Then DialogResult = DialogResult.Cancel
    End Sub
End Class

File: PolicyPlus/OpenPol.cs
using System;
using System.Windows.Forms;
using Microsoft.VisualBasic;

namespace PolicyPlus
{
    public partial class OpenPol
    {
        public PolicyLoader SelectedUser, SelectedComputer;

        public OpenPol()
        {
            InitializeComponent();
        }
        public void SetLastSources(PolicyLoaderSource ComputerType, string ComputerData, PolicyLoaderSource UserType, string UserData)
        {
            // Set up the UI to show the given configuration
            switch (ComputerType)
            {
                case PolicyLoaderSource.LocalGpo:
                    {
                        CompLocalOption.Checked = true;
                        break;
                    }
                case PolicyLoaderSource.LocalRegistry:
                    {
                        CompRegistryOption.Checked = true;
                        CompRegTextbox.Text = ComputerData;
                        break;
                    }
                case PolicyLoaderSource.PolFile:
                    {
                        CompFileOption.Checked = true;
                        CompPolFilenameTextbox.Text = ComputerData;
                        break;
                    }
                case PolicyLoaderSource.Null:
                    {
                        CompNullOption.Checked = true;
                        break;
                    }
            }
            switch (UserType)
            {
                case PolicyLoaderSource.LocalGpo:
                    {
                        UserLocalOption.Checked = true;
                        break;
                    }
                case PolicyLoaderSource.LocalRegistry:
                    {
                        UserRegistryOption.Checked = true;
                        UserRegTextbox.Text = UserData;
                        break;
                    }
                case PolicyLoaderSource.PolFile:
                    {
                        UserFileOption.Checked = true;
                        UserPolFilenameTextbox.Text = UserData;
                        break;
                    }
                case PolicyLoaderSource.SidGpo:
                    {
                        UserPerUserGpoOption.Checked = true;
                        UserGpoSidTextbox.Text = UserData;
                        break;
                    }
                case PolicyLoaderSource.NtUserDat:
                    {
                        UserPerUserRegOption.Checked = true;
                        UserHivePathTextbox.Text = UserData;
                        break;
                    }
                case PolicyLoaderSource.Null:
                    {
                        UserNullOption.Checked = true;
                        break;
                    }
            }
        }
        public void BrowseForPol(TextBox DestTextbox)
        {
            // Browse for a POL file and put it in a text box
            using (var sfd = new SaveFileDialog())
            {
                sfd.OverwritePrompt = false;
                sfd.Filter = "Registry policy files|*.pol";
                if (sfd.ShowDialog() == DialogResult.OK)
                    DestTextbox.Text = sfd.FileName;
            }
        }
        private void CompOptionsCheckedChanged(object sender, EventArgs e)
        {
            // When the user changes the computer selection
            bool regMount = CompRegistryOption.Checked;
            CompRegTextbox.Enabled = regMount;
            bool polActive = CompFileOption.Checked;
            CompPolFilenameTextbox.Enabled = polActive;
            CompFileBrowseButton.Enabled = polActive;
        }
        private void UserOptionsCheckedChanged(object sender, EventArgs e)
        {
            // When the user changes the user selection
            bool regMount = UserRegistryOption.Checked;
            UserRegTextbox.Enabled = regMount;
            bool file = UserFileOption.Checked;
            UserPolFilenameTextbox.Enabled = file;
            UserFileBrowseButton.Enabled = file;
            bool perUserGpo = UserPerUserGpoOption.Checked;
            UserGpoSidTextbox.Enabled = perUserGpo;
            UserBrowseGpoButton.Enabled = perUserGpo;
            bool perUserHive = UserPerUserRegOption.Checked;
            UserHivePathTextbox.Enabled = perUserHive;
            UserBrowseHiveButton.Enabled = perUserHive;
        }
        private void CompFileBrowseButton_Click(object sender, EventArgs e)
        {
            BrowseForPol(CompPolFilenameTextbox);
        }
        private void UserFileBrowseButton_Click(object sender, EventArgs e)
        {
            BrowseForPol(UserPolFilenameTextbox);
        }
        private void UserBrowseRegistryButton_Click(object sender, EventArgs e)
        {
            // Browse for a user Registry hive
            if (My.MyProject.Forms.OpenUserRegistry.ShowDialog() == DialogResult.OK)
                UserHivePathTextbox.Text = My.MyProject.Forms.OpenUserRegistry.SelectedFilePath;
        }
        private void OkButton_Click(object sender, EventArgs e)
        {
            // Create the policy loaders
            try
            {
                if (CompLocalOption.Checked)
                {
                    SelectedComputer = new PolicyLoader(PolicyLoaderSource.LocalGpo, "", false);
                }
                else if (CompRegistryOption.Checked)
                {
                    SelectedComputer = new PolicyLoader(PolicyLoaderSource.LocalRegistry, CompRegTextbox.Text, false);
                }
                else if (CompFileOption.Checked)
                {
                    SelectedComputer = new PolicyLoader(PolicyLoaderSource.PolFile, CompPolFilenameTextbox.Text, false);
                }
                else
                {
                    SelectedComputer = new PolicyLoader(PolicyLoaderSource.Null, "", false);
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox("The computer policy loader could not be created. " + ex.Message, MsgBoxStyle.Exclamation);
                return;
            }
            try
            {
                if (UserLocalOption.Checked)
                {
                    SelectedUser = new PolicyLoader(PolicyLoaderSource.LocalGpo, "", true);
                }
                else if (UserRegistryOption.Checked)
                {
                    SelectedUser = new PolicyLoader(PolicyLoaderSource.LocalRegistry, UserRegTextbox.Text, true);
                }
                else if (UserFileOption.Checked)
                {
                    SelectedUser = new PolicyLoader(PolicyLoaderSource.PolFile, UserPolFilenameTextbox.Text, true);
                }
                else if (UserPerUserGpoOption.Checked)
                {
                    SelectedUser = new PolicyLoader(PolicyLoaderSource.SidGpo, UserGpoSidTextbox.Text, true);
                }
                else if (UserPerUserRegOption.Checked)
                {
                    SelectedUser = new PolicyLoader(PolicyLoaderSource.NtUserDat, UserHivePathTextbox.Text, true);
                }
                else
                {
                    SelectedUser = new PolicyLoader(PolicyLoaderSource.Null, "", true);
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox("The user policy loader could not be created. " + ex.Message, MsgBoxStyle.Exclamation);
                return;
            }
            DialogResult = DialogResult.OK;
        }
        private void UserBrowseGpoButton_Click(object sender, EventArgs e)
        {
            // Browse for a per-user GPO
            if (My.MyProject.Forms.OpenUserGpo.ShowDialog() == DialogResult.OK)
                UserGpoSidTextbox.Text = My.MyProject.Forms.OpenUserGpo.SelectedSid;
        }
        private void OpenPol_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Escape)
                DialogResult = DialogResult.Cancel;
        }
    }
}